crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -1251,6 +1266,1158 @@ impl Display for IncorrectAddressLength {
+    )]
+    RollupIdsNotInSequencerBlock,
+    #[error(
+        "failed decoding an entry in the cometbft block.data field as a protobuf signed astria \
+         transaction"
+    )]
+    SignedTransactionProtobufDecode(#[source] prost::DecodeError),
+    #[error(
+        "failed converting a raw protobuf signed transaction decoded from the cometbft block.data
+        field to a native astria signed transaction"
+    )]
+    RawSignedTransactionConversion(#[source] SignedTransactionError),
+    #[error(
+        "the root derived from the rollup transactions in the cometbft block.data field did not \
+         match the root stored in the same block.data field"
+    )]
+    RollupTransactionsRootDoesNotMatchReconstructed,
+    #[error(
+        "the root derived from the rollup IDs in the cometbft block.data field did not match the \
+         root stored in the same block.data field"
+    )]
+    RollupIdsRootDoesNotMatchReconstructed,
+}
+
+/// A shadow of [`SequencerBlock`] with full public access to its fields.
+///
+/// This type does not guarantee any invariants and is mainly useful to get
+/// access the sequencer block's internal types.
+#[derive(Clone, Debug)]
+pub struct UncheckedSequencerBlock {
+    /// The original `CometBFT` header that was the input to this sequencer block.
+    pub header: tendermint::block::header::Header,
+    /// The collection of rollup transactions that were included in this block.
+    pub rollup_transactions: IndexMap<RollupId, Vec<Vec<u8>>>,
+    // The proof that the rollup transactions are included in the `CometBFT` block this
+    // sequencer block is derived form. This proof together with
+    // `Sha256(MTH(rollup_transactions))` must match `header.data_hash`.
+    // `MTH(rollup_transactions)` is the Merkle Tree Hash derived from the
+    // rollup transactions.
+    pub rollup_transactions_proof: merkle::Proof,
+    // The proof that the rollup IDs listed in `rollup_transactions` are included
+    // in the `CometBFT` block this sequencer block is derived form. This proof together
+    // with `Sha256(MTH(rollup_ids))` must match `header.data_hash`.
+    // `MTH(rollup_ids)` is the Merkle Tree Hash derived from the rollup IDs listed in
+    // the rollup transactions.
+    pub rollup_ids_proof: merkle::Proof,
+}
+
+/// `SequencerBlock` is constructed from a tendermint/cometbft block by
+/// converting its opaque `data` bytes into sequencer specific types.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SequencerBlock {
+    /// The result of hashing `header`. Guaranteed to not be `None` as compared to
+    /// the cometbft/tendermint-rs return type.
+    block_hash: [u8; 32],
+    /// The original `CometBFT` header that was the input to this sequencer block.
+    header: tendermint::block::header::Header,
+    /// The collection of rollup transactions that were included in this block.
+    rollup_transactions: IndexMap<RollupId, Vec<Vec<u8>>>,
```

Comment:
how does `IndexMap`'s ordering work compared to `BTreeMap`? assuming it's different, that's a breaking change

Reply:
This new grouping mechanism is used inside sequencer as well. You can see that the snapshot test is unaffected.

When the indexmap is constructed it is ordered by its keys after the fact. But this is the same order as that used for the btreemap.

Reply:
BtreeMap is balancing while it's being constructed. That gives nlogn overhead. Insertion the indexmap is linear and the final ordering has the same complexity.

Access on the other hand is O(1) where it is log n for the btreemap.

The indexmap is cache efficient, the btreemap is not.

So for our purpose the index map is strictly better.